### ISSUE TITLE
Persistent Incomplete Download

### ISSUE THREAD

**Issue by MrxHellsing** [Jules-WinnfieldX/CyberDropDownloader#58]
**Describe the bug**
The app downloads few megabytes and then show all downloads completed even if they aren't. Rerunning the app multiple times doesn't do anything to fix it. All incomplete downloads are shown as mp4 files with few byte size. There are no ".download" files.

**To Help Reproduce**
1. I am not sure if this is exclusive to bunkr.is files or not but it happened a day or two ago too but I didn't notice that and all the links were from bunkr.is I think and since the app showed the downloads as completed so didn't bother and removed the links from URL.txt but later found out those were actually incomplete downloads. Furthermore, today the same thing occurred.
2. Just downloaded and tried 2.3.4 and still nothing.

**Screenshots**
[https://i.imgur.com/CL5VIdQ.jpg]

**Setup Information**
 - OS: [Windows 10, 64-bit, 20H2 OS Build 19042.1526]
 - Cyberdrop Downloader Version #2.3.4

**Additional context**
[logs.log](https://github.com/Jules-WinnfieldX/CyberDropDownloader/files/8135558/logs.log)

**Comment by Jules-WinnfieldX** [comment]
This problem is usually indicative of network issues on the users end. Check for yourself if the links that were scraped (can be found in the logs.log file you uploaded) work in your browser. Delete the bad files, turn off any antivirus, vpn or proxy and try again.

**Comment by MrxHellsing** [comment]
I did what you said, but it seems like the problem lies in the download limit. Apparently, bunkr has a download limit for every IP address. I was indeed getting a network error on my end, even on the links that had been downloaded fully. After I turned on VPN and reloaded the same link, it immediately displayed correctly, and the app was downloading too. However, ran into another problem. I am not sure exactly what is the issue here but probably it has something to do with multiple files downloading simultaneously and this is quite random I think. When I downloaded a cyberdrop link, it downloaded about 10 files simultaneously. In bunkr's case, it downloaded 2-4 files at once, and maybe due to multiple downloads at the same time, it is running into download limit for all files or at least 3-4 or more files. How about adding a feature to limit the number of simultaneous downloads that most downloaders have. That way, if it does run into download limit, then at least it will be for that file and will not proceed to the next one before finishing the current one. This way, it will be easy to check which file it ran into limit instead of opening all files and checking if it had been downloaded fully or not.
.
Also, since the download is somehow IP bound, how about adding proxy support? Kind of like what Megabasterd has, even though that app doesn't work most of the time and not sure how that app works at all or if it even makes use of proxy at all caz you run the app, it downloads about 3 GB and then just stop, naturally, it should run the next proxy to continue downloading but it doesn't. Not sure if this will be feasible for your app or not, but I'm just putting out ideas.

**Comment by Jules-WinnfieldX** [comment]
Bunkr does not have a download limit. Bunkr only restricts to 2 simultaneous downloads. The program already restricts to 2 simultaneous downloads for bunkr either way. You are not hitting a limit with bunkr, you are having issues on your end with part of your configuration.

There already is a feature to restrict how many simultaneous downloads are taking place, and that's the threads setting in settings.py.

Addressed downloads above, they are IP bound, but not in the way you are intending. To re-iterate, you are not hitting a limit with bunkr.

**Comment by MrxHellsing** [comment]
Well then, how do I fix this problem that you're referring to that I have on my network configuration. I am at a loss here. Back then when I opened the auto-closing issue, after your fix, it downloaded everything well and didn't run into these kinds of problems, well at least not this often. Since then I didn't change anything in my configuration, it is what it was back then. Not sure, what is causing the issue. I don't run VPN unless I have to and since bunkr or cyberdrop isn't blocked here, so don't need to run VPN for that either. I don't run any proxy. My DNS is Cloudflare. I do run a Dynamic IP with Double NAT restriction though and because of that I run into seeding problems in Private torrent, other than that, nothing much that I know of.

**Comment by MrxHellsing** [comment]
Well, I download a lot on a daily basis from Mega, CyberDrop, Bunkr. Maybe that's why I am experiencing more issues. So far I haven't tried downloading from CyberDrop yet with your app. I'll try a few downloads from CyberDrop and see if the issue persists and let you know about it.

**Comment by MrxHellsing** [comment]
Update on CyberDrop site downloads: I'm currently downloading 2 links from cyberdrop and the total amount of videos is about 300+. So far 299 have been completed. I am seeing new downloads are being written as .download and all videos so far have been downloaded fully, not partially. The thread count is set to "1" to download 1 file at a time. After cyberdrop completes, I'll add 2 links from bunkr as well. As far as cyberdrop is concerned, I am not seeing any downloads incomplete as mp4 files.

**Comment by axel845845** [comment]
Hi, How can I update the script to the latest version? Sorry for my ignorance

**Comment by DeadManLogan** [comment]
Just run into the same problem. There are not any .download files left but most of the files (.mp4) are not opening. It probably has to do something with your network. If you try to download the files by clicking the "download media"  button on site, it will work perfectly. If not maybe the files are uploaded corrupted.

**Comment by Jules-WinnfieldX** [comment]
I've yet to be able to recreate this, any additional info would be handy. 

> Just run into the same problem. There are not any .download files left but most of the files (.mp4) are not opening. It probably has to do something with your network. If you try to download the files by clicking the "download media" button on site, it will work perfectly. If not maybe the files are uploaded corrupted.

**Comment by Jules-WinnfieldX** [comment]
A logs.log file from a run where this happens would probably be best. the logs.log file in the original post happened on a consequent run when the files already existed.

**Comment by MrxHellsing** [comment]
> It probably has to do something with your network. If you try to download the files by clicking the "download media" button on site, it will work perfectly. If not maybe the files are uploaded corrupted.

The download media button sometimes doesn't work either, normally my download manager catches the download but sometimes a pop-up opens but no action. When I am downloading a file from bunkr and at the same time if I try to open another link, I sometimes get a Network error.

**Comment by MrxHellsing** [comment]
UPDATE on BUNKR: Yeah, just as I suspected. This problem keeps occurring only with Bunkr files based on my observation. A lot of files are in mp4 files with few bytes and a vast majority of the files are full mp4 files with 60-70% data missing. I am attaching log file of today, the download was in progress from 2nd March, 2022 nighttime I think and finished today.
.
[logs.log](https://github.com/Jules-WinnfieldX/CyberDropDownloader/files/8185306/logs.log)

**Comment by Jules-WinnfieldX** [comment]
So two things. The first being I'm confused as for a vast majority of the files that you attempted to download, the files already existed on your computer (whether they were good or bad is another thing).

The second being, as you stated yesterday you said that the download media button often doesn't work for you on the website itself. These two things are going to be connected in the issue for you. If you open those few kb .mp4 files in a text editor you will see that they most likely contain html code (a webpage, not actual a video file). What this program does is grabs the links from that download button and just downloads them. If that button does not work for you, for whatever reason it may be, the program will also not function.

**Comment by MrxHellsing** [comment]
1) I am talking about the whole package, I didn't check until all 30+ GB was finished downloading. After checking, I found that a lot of the files are downloaded partially but showing as a full mp4 file and you can even play those but they only have contents of a few mins. For example: A video that is initially 10 mins, the downloaded file would only have about 2-3 or sometimes up to 8-9 mins, 1 min or more would be missing.
.
2) The download media button is kinda like a paradox, it works and then it doesn't. The timing is very random. If I am to give an example that would be this, suppose I am downloading a link that has about 20 files. Now, 5 of them completed fully but 15 of them failed ( by fail, I mean downloaded as a full mp4 file but has mins left out ). Now, at that very moment, if I open one of the failed files on the browser and attempt to download by clicking on it, it will open a pop-up but immediately close. I confirm that the button doesn't work by seeing my IDM pop up. If that doesn't have another full video listed below the 2KB mp4 file, then for that moment, the button wasn't working. Also, if I open multiple links at the same time, then in 2-3 tabs the button works and IDM captures the download, in the rest of the tabs, the buttons initiates a pop-up but nothing happens, and in those tabs, there is no full video listed below the 2KB mp4 file in IDM pop-up either. So, the working part is very random.

**Comment by Jules-WinnfieldX** [comment]
We are essentially talking about the same thing. With the retry function, if the download button is intermittent for you, the exact same result you are experiencing is going to happen. If the first attempt works for the first 9 minutes for example, and then the download stops and it has to try and resume, no the second attempt the button doesn't work, it's going to thing the download is complete. The header on the second attempt is going to be kb's in size, where you will have hundreds of mb's or gb's, it's going to drop the .download and assign it as a .mp4.

**Comment by Jules-WinnfieldX** [comment]
We use a resume point (range) in the header information, I have no idea what that would return in your case where the button isn't working. It could be 0, it could be something completely random, but not the rest of the file. Your issue isn't with my downloader, it's working as intended. Your issue is with Bunkr.

**Comment by MrxHellsing** [comment]
Yeah, I think that too, the problem is with Bunkr. It has some sort of limit in place that prevents a full download or prevents multiple download request or something similar to that. There are sites where download with IDM becomes impossible due to something called request being denied, basically, the browser sends a download request and at the same time IDM sends one too, so some sites are configured to deny the same request two in a row. I think bunkr has some sort of limit in place. Is there any way to bypass the limit through the app? Based on my observation, the app has no issue with CyberDrop as it has no limit in place.

**Comment by MitchPetersen** [comment]
I’ve talked to the owner of bunkr, there is no limit in place.

there is a simultaneous download limit of 2, which I already talked about above. Other than that there isn’t anything.

**Comment by Jules-WinnfieldX** [comment]
There is no limit in place besides the simultaneous download limit of 2. That is the only limit. If you are downloading with your browser or IDM while using my program you will run into issues. There is no other limit. I’ve talked to the owner and he confirmed that.
